File: PInvoke/Parser/Macro.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Imports System.Text.RegularExpressions

Namespace Parser


    ''' <summary>
    ''' Represents a macro in native code
    ''' </summary>
    ''' <remarks></remarks>
    <DebuggerDisplay("{Name} -> {Value}")> _
    Public Class Macro
        Private _name As String
        Private _value As String
        Private _isMethod As Boolean
        Private _isPermanent As Boolean
        Private _isFromParse As Boolean = True

        ''' <summary>
        ''' Name of the Macro
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Name() As String
            Get
                Return _name
            End Get
            Set(ByVal value As String)
                _name = value
            End Set
        End Property

        ''' <summary>
        ''' Value of the macro
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Value() As String
            Get
                Return _value
            End Get
            Set(ByVal val As String)
                _value = val
            End Set
        End Property

        ''' <summary>
        ''' Whether or not this is a method style macro
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property IsMethod() As Boolean
            Get
                Return _isMethod
            End Get
            Set(ByVal value As Boolean)
                _isMethod = value
            End Set
        End Property

        ''' <summary>
        ''' Represents a macro that cannot be overriden by user code.  
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property IsPermanent() As Boolean
            Get
                Return _isPermanent
            End Get
            Set(ByVal value As Boolean)
                _isPermanent = value
            End Set
        End Property

        ''' <summary>
        ''' Is this macro created from actually parsing code?  The alternate is that the 
        ''' macro is added to the initial set of macros.  This allows the parser to determine
        ''' what is actually a part of the parsed code as opposed to the setup code
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Friend Property IsFromParse() As Boolean
            Get
                Return _isFromParse
            End Get
            Set(ByVal value As Boolean)
                _isFromParse = value
            End Set
        End Property

        Public Sub New(ByVal name As String)
            _name = name
        End Sub

        Public Sub New(ByVal name As String, ByVal val As String)
            MyClass.New(name, val, False)
        End Sub

        Public Sub New(ByVal name As String, ByVal val As String, ByVal permanent As Boolean)
            _name = name
            _value = val
            _isPermanent = permanent
        End Sub

    End Class

    ''' <summary>
    ''' Macros that are methods
    ''' </summary>
    ''' <remarks></remarks>
    Public Class MethodMacro
        Inherits Macro

        Private _paramList As List(Of String)
        Private _bodyList As List(Of Token)
        Private _fullBodyList As List(Of Token)

        ''' <summary>
        ''' Text parameters of the macro
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Parameters() As List(Of String)
            Get
                Return _paramList
            End Get
        End Property

        ''' <summary>
        ''' Tokens inside the macro body minus any whitespace characters
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property Body() As List(Of Token)
            Get
                Return _bodyList
            End Get
        End Property

        ''' <summary>
        ''' Tokens inside the macro body including anywhitespace characters
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property FullBody() As List(Of Token)
            Get
                Return _fullBodyList
            End Get
        End Property

        ''' <summary>
        ''' Get the text of the method signature
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property MethodSignature() As String
            Get
                Dim b As New Text.StringBuilder()
                b.Append("(")
                For i As Integer = 0 To _paramList.Count - 1
                    b.Append(_paramList(i))
                    If i + 1 < _paramList.Count Then
                        b.Append(",")
                    End If
                Next
                b.Append(") ")
                For Each cur As Token In FullBody
                    b.Append(cur.Value)
                Next

                Return b.ToString()
            End Get
        End Property

        Public Sub New(ByVal name As String, ByVal paramList As List(Of String), ByVal body As List(Of Token), ByVal fullBody As List(Of Token))
            MyBase.New(name)
            MyBase.Value = name & "()"
            MyBase.IsMethod = True

            _paramList = paramList
            _bodyList = body
            _fullBodyList = fullBody
        End Sub

        Public Function Replace(ByVal argList As List(Of Token)) As List(Of Token)
            If argList.Count <> _paramList.Count Then
                Return New List(Of Token)
            End If

            ' Replace is done in 2 passes.  The first puts the arguments into the token stream.
            Dim retList As New List(Of Token)
            For Each item As Token In _bodyList
                If item.TokenType <> TokenType.Word Then
                    retList.Add(item)
                Else
                    Dim index As Int32 = _paramList.IndexOf(item.Value)
                    If index >= 0 Then
                        retList.Add(argList(index))
                    Else
                        retList.Add(item)
                    End If
                End If
            Next

            ' Second phase, process all of the # entries 
            Dim i As Int32 = 0
            While i < retList.Count - 1
                Dim curToken As Token = retList(i)
                Dim nextToken As Token = retList(i + 1)
                If curToken.TokenType = TokenType.Pound Then

                    If nextToken.TokenType = TokenType.Pound Then
                        ' Don't accidentally process a ## as a # token
                        i += 1
                    ElseIf argList.IndexOf(nextToken) >= 0 Then
                        If nextToken.IsQuotedString Then
                            ' Already quoted so it doesn't need to be quoted again
                            retList.RemoveAt(i)
                            i += 1
                        Else
                            ' Quote me macro
                            retList(i) = New Token(TokenType.QuotedStringAnsi, """" & nextToken.Value & """")
                            retList.RemoveAt(i + 1)
                        End If
                    End If
                End If

                i += 1
            End While

            Return retList
        End Function

        Public Shared Function TryCreateFromDeclaration(ByVal name As String, ByVal body As String, ByRef method As MethodMacro) As Boolean
            Try
                Dim engine As New PreProcessorEngine(New PreProcessorOptions)
                Using reader As New IO.StringReader("#define " & name & body)
                    engine.Process(New TextReaderBag(reader))
                    Dim created As Macro = Nothing
                    If engine.MacroMap.TryGetValue(name, created) AndAlso created.IsMethod Then
                        method = DirectCast(created, MethodMacro)
                        Return True
                    End If
                End Using
            Catch
                Return False
            End Try

            Return False
        End Function

    End Class
End Namespace

File: PInvoke.Core/Parser/Macro.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
// Copyright (c) Microsoft Corporation.  All rights reserved.
using System.Text.RegularExpressions;

namespace Parser
{


	/// <summary>
	/// Represents a macro in native code
	/// </summary>
	/// <remarks></remarks>
	[DebuggerDisplay("{Name} -> {Value}")]
	public class Macro
	{
		private string _name;
		private string _value;
		private bool _isMethod;
		private bool _isPermanent;

		private bool _isFromParse = true;
		/// <summary>
		/// Name of the Macro
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string Name {
			get { return _name; }
			set { _name = value; }
		}

		/// <summary>
		/// Value of the macro
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string Value {
			get { return _value; }
			set { _value = value; }
		}

		/// <summary>
		/// Whether or not this is a method style macro
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public bool IsMethod {
			get { return _isMethod; }
			set { _isMethod = value; }
		}

		/// <summary>
		/// Represents a macro that cannot be overriden by user code.  
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public bool IsPermanent {
			get { return _isPermanent; }
			set { _isPermanent = value; }
		}

		/// <summary>
		/// Is this macro created from actually parsing code?  The alternate is that the 
		/// macro is added to the initial set of macros.  This allows the parser to determine
		/// what is actually a part of the parsed code as opposed to the setup code
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		internal bool IsFromParse {
			get { return _isFromParse; }
			set { _isFromParse = value; }
		}

		public Macro(string name)
		{
			_name = name;
		}

		public Macro(string name, string val) : this(name, val, false)
		{
		}

		public Macro(string name, string val, bool permanent)
		{
			_name = name;
			_value = val;
			_isPermanent = permanent;
		}

	}

	/// <summary>
	/// Macros that are methods
	/// </summary>
	/// <remarks></remarks>
	public class MethodMacro : Macro
	{

		private List<string> _paramList;
		private List<Token> _bodyList;

		private List<Token> _fullBodyList;
		/// <summary>
		/// Text parameters of the macro
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public List<string> Parameters {
			get { return _paramList; }
		}

		/// <summary>
		/// Tokens inside the macro body minus any whitespace characters
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public List<Token> Body {
			get { return _bodyList; }
		}

		/// <summary>
		/// Tokens inside the macro body including anywhitespace characters
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public List<Token> FullBody {
			get { return _fullBodyList; }
		}

		/// <summary>
		/// Get the text of the method signature
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		public string MethodSignature {
			get {
				Text.StringBuilder b = new Text.StringBuilder();
				b.Append("(");
				for (int i = 0; i <= _paramList.Count - 1; i++) {
					b.Append(_paramList(i));
					if (i + 1 < _paramList.Count) {
						b.Append(",");
					}
				}
				b.Append(") ");
				foreach (Token cur in FullBody) {
					b.Append(cur.Value);
				}

				return b.ToString();
			}
		}

		public MethodMacro(string name, List<string> paramList, List<Token> body, List<Token> fullBody) : base(name)
		{
			base.Value = name + "()";
			base.IsMethod = true;

			_paramList = paramList;
			_bodyList = body;
			_fullBodyList = fullBody;
		}

		public List<Token> Replace(List<Token> argList)
		{
			if (argList.Count != _paramList.Count) {
				return new List<Token>();
			}

			// Replace is done in 2 passes.  The first puts the arguments into the token stream.
			List<Token> retList = new List<Token>();
			foreach (Token item in _bodyList) {
				if (item.TokenType != TokenType.Word) {
					retList.Add(item);
				} else {
					Int32 index = _paramList.IndexOf(item.Value);
					if (index >= 0) {
						retList.Add(argList(index));
					} else {
						retList.Add(item);
					}
				}
			}

			// Second phase, process all of the # entries 
			Int32 i = 0;
			while (i < retList.Count - 1) {
				Token curToken = retList(i);
				Token nextToken = retList(i + 1);

				if (curToken.TokenType == TokenType.Pound) {
					if (nextToken.TokenType == TokenType.Pound) {
						// Don't accidentally process a ## as a # token
						i += 1;
					} else if (argList.IndexOf(nextToken) >= 0) {
						if (nextToken.IsQuotedString) {
							// Already quoted so it doesn't need to be quoted again
							retList.RemoveAt(i);
							i += 1;
						} else {
							// Quote me macro
							retList(i) = new Token(TokenType.QuotedStringAnsi, "\"" + nextToken.Value + "\"");
							retList.RemoveAt(i + 1);
						}
					}
				}

				i += 1;
			}

			return retList;
		}

		public static bool TryCreateFromDeclaration(string name, string body, ref MethodMacro method)
		{
			try {
				PreProcessorEngine engine = new PreProcessorEngine(new PreProcessorOptions());
				using (IO.StringReader reader = new IO.StringReader("#define " + name + body)) {
					engine.Process(new TextReaderBag(reader));
					Macro created = null;
					if (engine.MacroMap.TryGetValue(name, created) && created.IsMethod) {
						method = (MethodMacro)created;
						return true;
					}
				}
			} catch {
				return false;
			}

			return false;
		}

	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
